File: apps/mobile/test/contracts/trade_math_contract_test.dart
// Contract tests for trade totals.
//
// These assert the shared fixtures in packages/contracts, which the JavaScript
// implementation in apps/web is held to as well. Cash is out of scope: this
// package takes it as an input, web derives it from the difference. See
// packages/contracts/README.md.
import 'package:flutter_test/flutter_test.dart';
import 'package:fabtrades/core/models/trade.dart';

import '../support/fixtures.dart';
import 'contract_fixtures.dart';

void main() {
  final contract = loadContract('trade_math');
  final tolerance = (contract['tolerance'] as num).toDouble();

  for (final c in contractCases(contract, 'cases')) {
    test(c['name'] as String, () {
      final trade = Trade(
        id: 'contract',
        createdAt: DateTime.utc(2026, 1, 1),
        haveItems: _items(c['have'], 'h'),
        wantItems: _items(c['want'], 'w'),
      );

      expect(
        trade.haveTotal,
        closeTo((c['haveTotal'] as num).toDouble(), tolerance),
      );
      expect(
        trade.wantTotal,
        closeTo((c['wantTotal'] as num).toDouble(), tolerance),
      );
      expect(trade.delta, closeTo((c['diff'] as num).toDouble(), tolerance));
    });
  }
}

List<TradeItem> _items(Object? raw, String idPrefix) {
  final list = (raw as List)
      .map((e) => Map<String, dynamic>.from(e as Map))
      .toList();
  return [
    for (var i = 0; i < list.length; i++)
      TradeItem(
        card: buildCard(id: '$idPrefix$i'),
        quantity: (list[i]['quantity'] as num).toInt(),
        priceEach: (list[i]['price'] as num).toDouble(),
      ),
  ];
}

File: apps/web/tests/contracts/tradeMath.contract.test.js
// Contract tests for trade totals.
//
// These assert the shared fixtures in packages/contracts, which the Dart
// implementation in apps/mobile is held to as well. Cash is out of scope: mobile
// takes it as an input, web derives it from the difference. See
// packages/contracts/README.md.
import contract from '../../../../packages/contracts/trade_math.json';
import { calculateTotal, calculateDiff } from '../../src/utils/trade.js';

describe('trade math contract', () => {
    contract.cases.forEach((testCase) => {
        it(testCase.name, () => {
            const haveTotal = calculateTotal(testCase.have);
            const wantTotal = calculateTotal(testCase.want);

            expect(haveTotal).toBeCloseTo(testCase.haveTotal, 9);
            expect(wantTotal).toBeCloseTo(testCase.wantTotal, 9);
            expect(calculateDiff(haveTotal, wantTotal)).toBeCloseTo(testCase.diff, 9);
        });
    });
});
